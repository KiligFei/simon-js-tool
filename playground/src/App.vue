<script setup lang="ts">
import { fileSplice } from '../../src'
// import { lazyLoad } from "../../src";
// import { vFetch } from "../../src";
// const upload = ref(null)
// onMounted(() => {
//   upload.value.addEventListener('change', (e) => {
//     const _file = upload.value.files[0]
//     fileSplice(_file)
//   })
// })
// setTimeout(() => {
//   vFetch({
//     url: "test1",
//   }).then((res: any) => {
//     // console.log(res)
//     return res;
//   });
// }, 2000);
// onMounted(() => {
//   lazyLoad(document.querySelectorAll("img"));
//   lazyLoad("img[data-src]");
//   lazyLoad(".img-wrapper>img");
//   // document.querySelectorAll("img").forEach((img) => {
//   //   lazyLoad(img);
//   // });
// });
</script>

<template>
  <main font-sans p="x-4 y-10" text="center gray-700 dark:gray-200">
    上传:<input ref="upload" type="file" name="" value="">
    <router-view />
    <Footer />
  </main>
  <!-- <img src="temp" data-src="../public/favicon.svg" alt="" h-10 bg-red /> -->
  <!-- <img src="temp" alt="" h-10 bg-red /> -->
</template>
