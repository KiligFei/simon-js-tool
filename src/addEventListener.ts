import { isStr } from './isStr'

export function addEventListener(target: Window | Document | Element | string, eventName: string, callback: (e: any) => void, useCapture?: boolean, autoRemove?: boolean): (() => void) {
  let mounted = false
  let stopped = false
  let stop: () => void
  if (eventName === 'DOMContentLoaded')
    stopped = true
  function event(e: Event) {
    callback.call(e.target, e)
    if (autoRemove)
      stop()
  }
  update()
  window.addEventListener('DOMContentLoaded', update)

  function update() {
    if (isStr(target))
      target = document.querySelector(target as string) as Element || target
    if (!mounted && isStr(target))
      return mounted = true
    else if (isStr(target))
      throw new Error(`${target} is not a Element`);
    (target as Element).addEventListener(eventName, event, useCapture)
    stop = () => (target as Element).removeEventListener(eventName, event, useCapture)
    if (stopped)
      stop()
  }
  return () => {
    if (!stop) {
      stopped = true
      return
    }
    stop?.()
  }
}

