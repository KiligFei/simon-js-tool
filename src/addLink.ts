import { createElement } from './createElement'
export function addLink(str: string) {
  document.head.appendChild(createElement('link', {
    rel: 'stylesheet',
    href: str,
    type: 'text/css',
  }))
}
