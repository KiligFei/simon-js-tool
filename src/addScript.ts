export function addScript(src: string) {
  try {
    const s = document.createElement('script')
    s.type = 'text/javascript'
    s.async = true
    s.src = src
    const t = document.getElementsByTagName('script')[0]
    t.parentNode?.insertBefore(s, t)
  }
  catch (error: any) {
    throw new Error(error)
  }
}
