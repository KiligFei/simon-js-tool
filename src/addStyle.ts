export function addStyle(s: string) {
  try {
    const style = document.createElement('style')
    style.innerHTML = s
    document.head.appendChild(style)
  }
  catch (error: any) {
    throw new Error(error)
  }
}
