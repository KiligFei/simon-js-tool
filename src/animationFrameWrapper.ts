export function animationFrameWrapper(fn: () => void, delta = 1000): (() => void) {
  let start: number; let work = true
  const animationFrame = window.requestAnimationFrame
    || window.webkitRequestAnimationFrame
    || window.mozRequestAnimationFrame
    || window.msRequestAnimationFrame
    || (fn => setTimeout(fn, 1000 / 60))
  const cancelAnimation = window.cancelAnimationFrame
    || window.webkitCancelAnimationFrame
    || window.mozCancelAnimationFrame
    || window.oCancelAnimationFrame
    || window.msCancelAnimationFrame
    || clearTimeout
  const animationId = animationFrame(function myFrame(timestamp: number = Date.now()) {
    if (!work)
      return
    if (start === undefined) { start = timestamp }
    else if (timestamp - start > delta) {
      fn?.()
      start = timestamp
    }
    animationFrame(myFrame)
  })
  return () => {
    work = false
    cancelAnimation(animationId)
  }
}
