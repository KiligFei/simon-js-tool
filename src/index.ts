export * from './types'
export { deepCompare } from './deepCompare'
export { deepMerge } from './deepMerge'
export { asyncPool } from './asyncPool'
