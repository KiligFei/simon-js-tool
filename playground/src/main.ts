import { createApp } from 'vue'
import { createRouter, createWebHistory } from 'vue-router'
import routes from 'virtual:generated-pages'
import { vFetch } from '../../src'
import App from './App.vue'

import '@unocss/reset/tailwind.css'
import './styles/main.css'
import 'uno.css'

const app = createApp(App)
const router = createRouter({
  history: createWebHistory(import.meta.env.BASE_URL),
  routes,
})
app.use(router)
app.mount('#app')

vFetch.interceptors.request.use((response) => {
  console.log(response)
  return response
}, (err) => {
  // console.log('err', err)
  return Promise.reject(err)
})
const instance = vFetch.create({
  baseURL: 'http://localhost:5001/',
})

instance({
  url: 'test',
}).then((res: any) => {
  // console.log(res)
  return res
})

// vFetch.get({
//   url: "http://localhost:5001/t"
// }).then(res => {
//   console.log(res)
// }, err => {
//   console.log(err)
// })
