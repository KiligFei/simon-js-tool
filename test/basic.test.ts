import { describe, expect, it } from 'vitest'
import { asyncPool, curry, debounce, deepClone, deepCompare, deepMerge, memorizeFn, quickFilter, quickFind, throttle, transformKey, traverse, uniqueArray } from '../src'

describe('Test 1', () => {
  it('deepMerge test', () => {
    const b = {
      people: {
        name: 'xx',
        age: '11',
        fn: {},
        arr: [1, 2, 3],
      },
      name1: {
        name2: '',
      },
    }
    const c = {
      people: {
        name: 'xx',
        age: '11',
        fn: {
          name: '11',
        },
        arr: [1, 2, 4],
      },
      name1: {
        name2: {},
      },
    }
    expect(deepMerge({}, b, c)).toMatchInlineSnapshot(`
      {
        "name1": {
          "name2": {},
        },
        "people": {
          "age": "11",
          "arr": [
            1,
            2,
            4,
          ],
          "fn": {
            "name": "11",
          },
          "name": "xx",
        },
      }
    `)
  })
})

describe('Test 2', () => {
  it('deepCompare test', () => {
    const b = {
      people: {
        name: 'xx',
        age: '11',
        fn: {},
        arr: [1, 2, 3, 5],
      },
      name1: {
        name2: '',
      },
    }
    const c = {
      people: {
        name: 'xx',
        age: '11',
        fn: {
          name: '11',
        },
        arr: [1, 2, 4],
      },
      name1: {
        name2: {},
      },
    }
    expect(deepCompare(b, c)).toMatchInlineSnapshot(`
      {
        "error": [
          "people.fn.name",
          "people.arr",
          "people.arr",
          "name1.name2",
        ],
        "errorMsg": [
          "people.fn.name数据不一致，分别为undefined   =>    11",
          "people.arr数据不一致，第2项,分别为3   =>    4",
          "people.arr数据不一致，第3项,分别为5   =>    undefined",
          "name1.name2数据不一致，分别为   =>    [object Object]",
        ],
      }
    `)

    const d = [1, 2, 3]
    const e = [1, 2, 4]
    expect(deepCompare(d, e)).toMatchInlineSnapshot(`
      {
        "error": [
          "array",
        ],
        "errorMsg": [
          "array数据不一致，第2项,分别为3   =>    4",
        ],
      }
    `)
  })
})

describe('Test 3', () => {
  it('asyncPool test', async () => {
    function delay(interval: number) {
      return () => new Promise((resolve, reject) => {
        setTimeout(() => {
          if (interval > 1000)
            reject(interval)
          else
            resolve(interval)
        }, interval)
      })
    }

    const tasks = [
      delay(1000),
      delay(1000),
      delay(1000),
      delay(1000),
      delay(1000),
      delay(1000),
      delay(1000),
      delay(1000),
    ]
    expect(await asyncPool(4, tasks)).toMatchInlineSnapshot(`
      [
        1000,
        1000,
        1000,
        1000,
        1000,
        1000,
        1000,
        1000,
      ]
    `)
  })
})

describe('Test 4', () => {
  it('quickFind test', async () => {
    const arr = [
      {
        name: 'simon',
        age: 18,
        id: 0,
      },
      {
        name: 'simon1',
        age: 19,
        id: 1,
      },
      {
        name: 'simon2',
        age: 29,
        id: 2,
      },
    ]
    const finder = quickFind(arr, 'id')
    expect(finder.find(2)).toMatchInlineSnapshot(`
      {
        "age": 29,
        "id": 2,
        "name": "simon2",
      }
    `)
    expect(finder.set(2, 'name', 'new value~~')).toMatchInlineSnapshot(`
      [
        {
          "age": 18,
          "id": 0,
          "name": "simon",
        },
        {
          "age": 19,
          "id": 1,
          "name": "simon1",
        },
        {
          "age": 29,
          "id": 2,
          "name": "new value~~",
        },
      ]
    `)
    expect(finder.set(3, { id: 3, name: 'new value' })).toMatchInlineSnapshot(`
      [
        {
          "age": 18,
          "id": 0,
          "name": "simon",
        },
        {
          "age": 19,
          "id": 1,
          "name": "simon1",
        },
        {
          "age": 29,
          "id": 2,
          "name": "new value~~",
        },
        {
          "id": 3,
          "name": "new value",
        },
      ]
    `)
    expect(finder.delete(3)).toMatchInlineSnapshot(`
      [
        {
          "age": 18,
          "id": 0,
          "name": "simon",
        },
        {
          "age": 19,
          "id": 1,
          "name": "simon1",
        },
        {
          "age": 29,
          "id": 2,
          "name": "new value~~",
        },
      ]
    `)
  })
})

describe('Test 5', () => {
  it('quickFilter test', async () => {
    const arr = [
      {
        name: 'simon',
        age: 18,
        id: 0,
      },
      {
        name: 'simon1',
        age: 19,
        id: 1,
      },
      {
        name: 'simon2',
        age: 29,
        id: 2,
      },
      {
        name: 'simon3',
        age: 29,
        id: 20,
      },
      {
        name: 'simon3',
        age: 39,
        id: 22,
      },
      {
        name: 'simon5',
        age: 49,
        id: 3,
      },
    ]
    expect(quickFilter(arr, ['age', 'id'], '3')).toMatchInlineSnapshot(`
      [
        {
          "age": 39,
          "id": 22,
          "name": "simon3",
        },
        {
          "age": 49,
          "id": 3,
          "name": "simon5",
        },
      ]
    `)
  })
})

describe('Test 6', () => {
  it('deepClone test', async () => {
    const arr: any = {
      name: 'simon5',
      age: 49,
      id: 3,
      fn() { },
      reg: /\test/,
      date: Date,
      null: null,
      [Symbol(1)]: 1,
      loop: null,
    }
    arr.loop = arr
    const cloneObj = deepClone(arr)
    expect(cloneObj).toMatchInlineSnapshot('false')
    expect(cloneObj.loop === arr.loop).toMatchInlineSnapshot('10')
  })
})

describe('Test 7', () => {
  it('curry test', async () => {
    function __add(a: number, b: number, c: number, d: number) {
      return a + b + c + d
    }
    const add = curry(__add)
    expect(add(1, 2)(3)(4)).toBe(10)
  })
})

describe('Test 8', () => {
  it('memorizeFn test', async () => {
    let count = 0
    function hello(name: string) {
      count++
      return `hello ${name}`
    }
    const fn = memorizeFn(hello)
    fn()
    fn()
    fn()
    expect(count).toBe(1)
  })
})

describe('Test 9', () => {
  it('traverse test', async () => {
    const arr = [
      {
        name: 'simon',
        family: {
          bro: 'simonBro',
        },
        people: {
          name: 'simonPeople',
          family: {
            bro: 'simonPeopleBro',
          },
        },
      }, {
        name: 'kitty',
        family: {
          bro: 'kittyBro',
        },
        people: {
          name: 'kittyPeople',
          family: {
            bro: 'kittyPeopleBro',
          },
        },
      },
    ]
    expect(traverse(arr, {
      'family.bro': function (target: any, index: number) {
        console.log('traverse~', target, index)
      },
      people(target: any, index: number) {
        target.name = 'haha'
        console.log('traverse~', target, index)
      },
      'people.family': function (target: any, index: number, item: any) {
        console.log('traverse~', target, index, item)
      },
    })).toMatchInlineSnapshot(`
      [
        {
          "family": {
            "bro": "simonBro",
          },
          "name": "simon",
          "people": {
            "family": {
              "bro": "simonPeopleBro",
            },
            "name": "haha",
          },
        },
        {
          "family": {
            "bro": "kittyBro",
          },
          "name": "kitty",
          "people": {
            "family": {
              "bro": "kittyPeopleBro",
            },
            "name": "haha",
          },
        },
      ]
    `)
  })
})

describe('Test 10', () => {
  it('transformKey test', async () => {
    const arr = [
      {
        name: 'simon',
        family: {
          bro: 'simonBro',
        },
        people: {
          name: 'simonPeople',
          family: {
            bro: 'simonPeopleBro',
          },
        },
      },
    ]
    expect(transformKey(arr, {
      'family.bro': 'name',
      'people.family.bro': 'familyName',
    })).toMatchInlineSnapshot(`
      [
        {
          "family": {
            "name": "simonBro",
          },
          "name": "simon",
          "people": {
            "family": {
              "familyName": "simonPeopleBro",
            },
            "name": "simonPeople",
          },
        },
      ]
    `)
  })
})

describe('Test 11', () => {
  it('debounce test', async () => {
    let count = 0
    function add() {
      count++
    }
    const fn = debounce(add, 1000)
    fn()
    fn()
    fn()
    fn()
    fn()
    setTimeout(() => {
      expect(count).toBe(1)
    }, 1000)
  })
})

describe('Test 12', () => {
  it('throttle test', async () => {
    let count = 0
    function add() {
      count++
    }
    const fn = throttle(add, 1000)
    fn()
    fn()
    fn()
    expect(count).toBe(1)
  })
})

describe('Test 13', () => {
  it('uniqueArray test', async () => {
    const array = [
      {
        name: 'simon',
        age: '18',
      },
      {
        name: 'simon',
        age: '18',
        hobby: 'basketball',

      },
      {
        name: 'simon',
        age: 19,
      },
      {
        name: 'simon',
        age: '19',
        hobby: ['1', '2', '3'],
      },
      {
        name: 'simon',
        age: '19',
        hobby: ['1', '2', '3'],
      },
      {
        name: 'simon',
        age: '19',
      },

    ]
    expect(uniqueArray(array)).toMatchInlineSnapshot(`
      [
        {
          "age": "18",
          "name": "simon",
        },
        {
          "age": "18",
          "hobby": "basketball",
          "name": "simon",
        },
        {
          "age": 19,
          "name": "simon",
        },
        {
          "age": "19",
          "hobby": [
            "1",
            "2",
            "3",
          ],
          "name": "simon",
        },
        {
          "age": "19",
          "name": "simon",
        },
      ]
    `)
  })
})
