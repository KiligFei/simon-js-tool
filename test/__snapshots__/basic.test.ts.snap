// Vitest Snapshot v1

exports[`Test htmlTransform > htmlTransform test 1`] = `"<div :class=\\"myclass\\" name=\\"simon\\"></div>"`;
